### PR TITLE
lib/ukrandom: Add README.md

### PR DESCRIPTION
**Pull request by michpappas**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Description of changes

Add documentation on `lib/ukrandom` and `drivers/ukrandom/lcpu`.

### CODE REVIEW

File: drivers/ukrandom/lcpu/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# CPU-generated Randomness
+
+This library implements a Unikraft driver that provides CPU-generated randomness via the [libukrandom driver API]( https://github.com/unikraft/unikraft/blob/staging/lib/ukrandom/include/uk/random/driver.h).
```
Comment (StefanJum) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
This library implements a Unikraft driver that provides CPU-generated randomness via the [libukrandom driver API](https://github.com/unikraft/unikraft/blob/staging/lib/ukrandom/include/uk/random/driver.h).
```

File: drivers/ukrandom/lcpu/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+Arm introduces secure randomness on `Armv8.5-A` via the `RNDR` / `RNDRRS` instructions, on processors that implement `FEAT_RNG`. For more info, see the [Arm Architecture Reference Manual](https://developer.arm.com/documentation/ddi0487).
```
Comment (StefanJum) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
Arm introduces secure randomness on `Armv8.5-A` via the `RNDR` / `RNDRRS` instructions, on processors that implement `FEAT_RNG`.
For more info, see the [Arm Architecture Reference Manual](https://developer.arm.com/documentation/ddi0487).
```

File: drivers/ukrandom/lcpu/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+This driver implements `random_bytes()` using `RNDR` and `seed_bytes()` using `RNDRRS`. `seed_bytes_fb()` falls back to `RNDR`, if `RNDRRS` fails to provide entropy during the call.
```
Comment (StefanJum) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
This driver implements `random_bytes()` using `RNDR` and `seed_bytes()` using `RNDRRS`.
`seed_bytes_fb()` falls back to `RNDR`, if `RNDRRS` fails to provide entropy during the call.
```

File: drivers/ukrandom/lcpu/README.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+Intel provides cryptographically secure randomness via `RDRAND` / `RDSEED`. `RDRAND` was introduced at 2012 with Intel Ivy Bridge / AMD Bulldozer, and `rdseed` was first part of Intel Broadwell / AMD Zen.
```
Comment (StefanJum) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
Intel provides cryptographically secure randomness via `RDRAND` / `RDSEED`.
`RDRAND` was introduced at 2012 with Intel Ivy Bridge / AMD Bulldozer, and `rdseed` was first part of Intel Broadwell / AMD Zen.
```

File: drivers/ukrandom/lcpu/README.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+Information on `RDRAND` / `RDSEED` is provided at Intel's [Digital Random Number Generator (DRNG) Software Implementation Guide](https://www.intel.com/content/www/us/en/developer/articles/guide/intel-digital-random-number-generator-drng-software-implementation-guide.html).
+
+This driver implements `random_bytes()` using `RDRAND`, and `seed_bytes()` using `RDSEED`.
+If `RDSEED` is not detected during probe, `seed_bytes()` falls-back to `RDRAND`. Similarly, `seed_bytes_fb()` falls back to `RDRAND`, if `RDSEED` is not detected, or if `RDSEED` fails to provide entropy during the call.
```
Comment (StefanJum) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
If `RDSEED` is not detected during probe, `seed_bytes()` falls-back to `RDRAND`.
Similarly, `seed_bytes_fb()` falls back to `RDRAND`, if `RDSEED` is not detected, or if `RDSEED` fails to provide entropy during the call.
```

File: drivers/ukrandom/lcpu/README.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+* **x86_64**: `cat /proc/cpuinfo | grep rdrand` and `cat /proc/cpuinfo | grep rdseed`
+
+### QEMU with emulation (TCG)
+
+Make sure you use a version of QEMU that implements CPU-provided randomness on TCG:
+
+* `arm64`: QEMU introduces `RNDR` / `RNDRRS` on TCG in [QEMU 4.1](https://wiki.qemu.org/ChangeLog/4.1)
+* `x86_64`: QEMU introduces `RDRAND` support on TCG in [QEMU 4.1](https://wiki.qemu.org/ChangeLog/4.1), and `RDSEED` in [QEMU 8.1](https://wiki.qemu.org/ChangeLog/8.1).
+
+You should also select a suitable CPU on QEMU. Passing `--cpu=max` is a safe choice.
```
Comment (StefanJum) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
You should also select a suitable CPU on QEMU.
Passing `--cpu=max` is a safe choice.
```

File: lib/ukrandom/README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
+# libukrandom: Cryptographically Secure Randomness
+
+This library implements a cryptographically-secure PRNG (CSPRNG).
+Secure randomness is available to applications via the `get_random()` syscall (`CONFIG_LIBUKRANDOM_GETRANDOM`), and via `/dev/random`, `/dev/urandom`, and `/dev/hwrng` (`CONFIG_LIBUKRANDOM_DEVFS`). Unikraft libraries can obtain secure randomness via the [libukrandom API](https://github.com/unikraft/unikraft/blob/staging/lib/ukrandom/include/uk/random.h).
```
Comment (StefanJum) [inline lib/ukrandom/README.md]:
```suggestion
Secure randomness is available to applications via the `get_random()` syscall (`CONFIG_LIBUKRANDOM_GETRANDOM`), and via `/dev/random`, `/dev/urandom`, and `/dev/hwrng` (`CONFIG_LIBUKRANDOM_DEVFS`).
Unikraft libraries can obtain secure randomness via the [libukrandom API](https://github.com/unikraft/unikraft/blob/staging/lib/ukrandom/include/uk/random.h).
```

File: lib/ukrandom/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+* Pass the `random.seed` option to the command-line, with a 256-bit seed formatted as an array of 32-bit words.
+
+To provide a seed via the device-tree:
+* Select `CONFIG_LIBUKRANDOM_DTB_SEED`
+* Pass a 256-bit seed formatted as an array of 32-bit words to the `/chosen/rng-seed` node.
+
+If you are populating the seed on the host, you can generate secure randomness via `/dev/urandom`.
+
+**Example:** Passing a seed via the command-line on QEMU
+```
```
Comment (StefanJum) [inline lib/ukrandom/README.md]:
```suggestion

```text
```

File: lib/ukrandom/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+SEED=$(hexdump -vn32 -e'8/4 "0x%08X "' /dev/urandom)
+```
+Then invoke QEMU as usual, passing the generated seed to the command line:
```
Comment (StefanJum) [inline lib/ukrandom/README.md:35]:
```suggestion

Then invoke QEMU as usual, passing the generated seed to the command line:
```

File: lib/ukrandom/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+```
```
Comment (StefanJum) [inline lib/ukrandom/README.md]:
```suggestion

```text
```

File: lib/ukrandom/README.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+-append "random.seed=[${SEED}] --"
+```
+
+**Notice:**
+* The command-line and dtb options have higher precedence over seeding via a hardware TRNG.
+* The command-line option involves secuirty risks (e.g. the seed is available in the host's procfs / ps). It is the responsibility of the loader / host to protect the seed.
```
Comment (StefanJum) [inline lib/ukrandom/README.md]:
```suggestion
* The command-line option involves secuirty risks (e.g. the seed is available in the host's procfs / ps).
  It is the responsibility of the loader / host to protect the seed.
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
Looks good @michpappas, just a few styling comments.

**Comment by michpappas** [comment]
@StefanJum thanks for the review. All comments addressed according to your suggestions.

File: drivers/ukrandom/lcpu/README.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+This driver implements `random_bytes()` using `RDRAND`, and `seed_bytes()` using `RDSEED`.
+If `RDSEED` is not detected during probe, `seed_bytes()` falls-back to `RDRAND`.
+Similarly, `seed_bytes_fb()` falls back to `RDRAND`, if `RDSEED` is not detected, or if `RDSEED` fails to provide entropy during the call.
+
+### QEMU with KVM acceleration / Firecracker / Baremetal
+
+To determine whether the host CPU provides randomness you can check `/proc/cpuinfo`:
+* **arm64**: `cat /proc/cpuinfo | grep rng`. Alternatively, you can refer to Marcin Juszkiewicz's [list of AArch64 SoC Features](https://gpages.juszkiewicz.com.pl/arm-socs-table/arm-socs.html).
+* **x86_64**: `cat /proc/cpuinfo | grep rdrand` and `cat /proc/cpuinfo | grep rdseed`
```
Comment (Mihnea0Firoiu) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
To determine whether the host CPU provides randomness you can check `/proc/cpuinfo`:

* **arm64**: `cat /proc/cpuinfo | grep rng`. Alternatively, you can refer to Marcin Juszkiewicz's [list of AArch64 SoC Features](https://gpages.juszkiewicz.com.pl/arm-socs-table/arm-socs.html).
* **x86_64**: `cat /proc/cpuinfo | grep rdrand` and `cat /proc/cpuinfo | grep rdseed`
```

File: lib/ukrandom/README.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+## Seeding the CSPRNG
+
+The PRNG is normally seeded by a hardware TRNG.
+Currently this is provided by `drivers/ukrandom/lcpu/` which provides CPU-generated randomness.
+
+On systems that don't feature a hardware TRNG, the host / loader can alternatively provide a seed via Unikraft's command-line, or on `arm64` systems via the device-tree.
+
+To provide a seed via the command line:
+* Select `CONFIG_LIBUKRANDOM_CMDLINE_SEED`
+* Pass the `random.seed` option to the command-line, with a 256-bit seed formatted as an array of 32-bit words.
```
Comment (Mihnea0Firoiu) [inline lib/ukrandom/README.md]:
```suggestion
To provide a seed via the command line:

* Select `CONFIG_LIBUKRANDOM_CMDLINE_SEED`
* Pass the `random.seed` option to the command-line, with a 256-bit seed formatted as an array of 32-bit words.
```

File: lib/ukrandom/README.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+To provide a seed via the device-tree:
+* Select `CONFIG_LIBUKRANDOM_DTB_SEED`
+* Pass a 256-bit seed formatted as an array of 32-bit words to the `/chosen/rng-seed` node.
```
Comment (Mihnea0Firoiu) [inline lib/ukrandom/README.md]:
```suggestion
To provide a seed via the device-tree:

* Select `CONFIG_LIBUKRANDOM_DTB_SEED`
* Pass a 256-bit seed formatted as an array of 32-bit words to the `/chosen/rng-seed` node.
```

File: lib/ukrandom/README.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+Then invoke QEMU as usual, passing the generated seed to the command line:
+
+```text
+-append "random.seed=[${SEED}] --"
+```
+
+**Notice:**
+* The command-line and dtb options have higher precedence over seeding via a hardware TRNG.
+* The command-line option involves secuirty risks (e.g. the seed is available in the host's procfs / ps).
+  It is the responsibility of the loader / host to protect the seed.
```
Comment (Mihnea0Firoiu) [inline lib/ukrandom/README.md:44]:
```suggestion
**Notice:**

* The command-line and dtb options have higher precedence over seeding via a hardware TRNG.
* The command-line option involves secuirty risks (e.g. the seed is available in the host's procfs / ps).
  It is the responsibility of the loader / host to protect the seed.
```

**Review by Mihnea0Firoiu — COMMENTED** [review]
Looks good, but shouldn't be blank lines here?

**Comment by michpappas** [comment]
> Looks good, but shouldn't be blank lines here?

@Mihnea0Firoiu, not sure whether this is based on a convention we have, or from an aesthetics point of view. For the latter, I did a render and it looks fine. Screenshot below:

image

**Comment by razvand** [comment]
> > Looks good, but shouldn't be blank lines here?
> 
> @Mihnea0Firoiu, not sure whether this is based on a convention we have, or from an aesthetics point of view. For the latter, I did a render and it looks fine. Screenshot below:

Hi, @michpappas . Yes, we have a convention to separate lists from the rest of the text with empty lines. It's something we use for the [`docs` repository](https://github.com/unikraft/docs). Rendering is the same, but it makes viewing / editing text easier.

File: drivers/ukrandom/lcpu/README.md [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+Information on `RDRAND` / `RDSEED` is provided at Intel's [Digital Random Number Generator (DRNG) Software Implementation Guide](https://www.intel.com/content/www/us/en/developer/articles/guide/intel-digital-random-number-generator-drng-software-implementation-guide.html).
+
+This driver implements `random_bytes()` using `RDRAND`, and `seed_bytes()` using `RDSEED`.
+If `RDSEED` is not detected during probe, `seed_bytes()` falls-back to `RDRAND`.
+Similarly, `seed_bytes_fb()` falls back to `RDRAND`, if `RDSEED` is not detected, or if `RDSEED` fails to provide entropy during the call.
+
+### QEMU with KVM acceleration / Firecracker / Baremetal
+
+To determine whether the host CPU provides randomness you can check `/proc/cpuinfo`:
+* **arm64**: `cat /proc/cpuinfo | grep rng`. Alternatively, you can refer to Marcin Juszkiewicz's [list of AArch64 SoC Features](https://gpages.juszkiewicz.com.pl/arm-socs-table/arm-socs.html).
```
Comment (razvand) [inline drivers/ukrandom/lcpu/README.md]:
```suggestion
* **arm64**: `cat /proc/cpuinfo | grep rng`.
  Alternatively, you can refer to Marcin Juszkiewicz's [list of AArch64 SoC Features](https://gpages.juszkiewicz.com.pl/arm-socs-table/arm-socs.html).
```

**Review by razvand — COMMENTED** [review]
Just a small comment. We prefer to have one sentence per line.

**Comment by michpappas** [comment]
> > > Looks good, but shouldn't be blank lines here?
> > 
> > 
> > @Mihnea0Firoiu, not sure whether this is based on a convention we have, or from an aesthetics point of view. For the latter, I did a render and it looks fine. Screenshot below:
> 
> Hi, @michpappas . Yes, we have a convention to separate lists from the rest of the text with empty lines. It's something we use for the [`docs` repository](https://github.com/unikraft/docs). Rendering is the same, but it makes viewing / editing text easier.

Okay I wasn't aware of that, thanks for clarifying. I pushed an updated version and indeed they render the same.

File: drivers/ukrandom/lcpu/README.md (hunk 16 above)
Comment (michpappas) [reply drivers/ukrandom/lcpu/README.md]: Done

File: lib/ukrandom/README.md [hunk 17]
```diff
@@ -0,0 +1,44 @@
+# libukrandom: Cryptographically Secure Randomness
+
+This library implements a cryptographically-secure PRNG (CSPRNG).
+Secure randomness is available to applications via the `get_random()` syscall (`CONFIG_LIBUKRANDOM_GETRANDOM`), and via `/dev/random`, `/dev/urandom`, and `/dev/hwrng` (`CONFIG_LIBUKRANDOM_DEVFS`).
```
Comment (razvand) [inline lib/ukrandom/README.md]:
```suggestion
Secure randomness is available to applications via the `getrandom()` syscall (`CONFIG_LIBUKRANDOM_GETRANDOM`), and via `/dev/random`, `/dev/urandom`, and `/dev/hwrng` (`CONFIG_LIBUKRANDOM_DEVFS`).
```

Isn't the system call [`getrandom()`](https://man7.org/linux/man-pages/man2/getrandom.2.html), with the underline?

**Review by razvand — CHANGES_REQUESTED** [review]
Just a small comment.

**Review by StefanJum — APPROVED** [review]
All good besides the comment from @razvand 
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

File: lib/ukrandom/README.md (hunk 17 above)
Comment (michpappas) [reply lib/ukrandom/README.md]: Good catch, thanks for spotting it out. I pushed a fix.

**Comment by michpappas** [comment]
@razvand this is now updated

**Review by razvand — APPROVED** [review]
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**